apps/prairielearn/src/tests/workspaceDynamicFiles.test.ts

```diff
@@ -0,0 +1,174 @@
+import * as fs from 'node:fs/promises';
+import { join } from 'node:path';
+
+import { assert } from 'chai';
+import * as cheerio from 'cheerio';
+import fetch from 'node-fetch';
+
+import { loadSqlEquiv, queryRow, queryRows } from '@prairielearn/postgres';
+
+import { config } from '../lib/config.js';
+import { IdSchema, IssueSchema } from '../lib/db-types.js';
+import { initialize } from '../lib/workspace.js';
+
+import * as helperServer from './helperServer.js';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const siteUrl = 'http://localhost:' + config.serverPort;
+const baseUrl = siteUrl + '/pl';
+
+const originalConfig = { ...config };
+
+let workspaceId: string | undefined;
+let sourcePath: string;
+let variantId: string | undefined;
+
+async function checkFileContents(filename: string, expectedContents: string) {
+  const fileContents = await fs.readFile(join(sourcePath, filename), 'binary');
+  assert.equal(fileContents, expectedContents);
+}
+
+async function createWorkspaceVariant(qid: string) {
+  const questionId = await queryRow(sql.select_test_question, { qid }, IdSchema);
+  const workspaceQuestionUrl = `${baseUrl}/course_instance/1/instructor/question/${questionId}/preview`;
+  const response = await fetch(workspaceQuestionUrl);
+
+  const $ = cheerio.load(await response.text());
+  const workspaceButton = $('a:contains("Open workspace")');
+  assert.lengthOf(workspaceButton, 1);
+
+  workspaceId = workspaceButton.attr('href')?.match('/pl/workspace/([0-9]+)')?.[1];
+  assert.isDefined(workspaceId);
+
+  variantId = $('.question-form input[name="__variant_id"]').attr('value');
+  assert.isDefined(variantId);
+}
+
+async function initializeWorkspace() {
+  assert.isDefined(workspaceId);
+  // A workspace is typically initialized using sockets, but to simplify the
+  // testing environment, we will call the initialization process directly.
+  ({ sourcePath } = await initialize(workspaceId));
+  const directoryStats = await fs.stat(sourcePath);
+  assert.isTrue(directoryStats.isDirectory());
+}
+
+describe('Test workspace dynamic files', function () {
+  this.timeout(20000);
+
+  before('set configuration values', async () => {
+    config.workspaceEnable = false;
+    config.workspaceHomeDirRoot = await fs.mkdtemp('/tmp/prairielearn-workspace-test-');
+  });
+  after('restore configuration values', async () => {
+    await fs.rm(config.workspaceHomeDirRoot, { force: true, recursive: true });
+    config.workspaceEnable = originalConfig.workspaceEnable;
+    config.workspaceHomeDirRoot = originalConfig.workspaceHomeDirRoot;
+  });
+  before('set up testing server', helperServer.before());
+  after('shut down testing server', helperServer.after);
+
+  describe('Question with valid dynamic files', () => {
+    it('create workspace question variant', async () => createWorkspaceVariant('workspace'));
+    it('initialize workspace', initializeWorkspace);
```

Comment:
Nit, it looks like these things are always done together; maybe we just a single `createAndInitializeWorkspace`?

Reply:
Done in d682c5b.